### PR TITLE
Ge suspendphysics

### PR DESCRIPTION
**Pull request by youle31**
test file: http://www.pasteall.org/blend/43239

### CODE REVIEW

**Comment by youle31** [comment]
This file: http://pasteall.org/blend/index.php?id=43227 shows that there is no culling for physics ?!

This is an Enormous performances loose. I think we should find a way to implement physics culling just by deleting physics controller on some conditions. The process could be partly controlled by the user. I add this to the framapad.

**Comment by youle31** [comment]
@panzergame : Do you think we have a risk with this patch? With children objects or other things?

**Comment by BluePrintRandom** [comment]
Is this patch ready?

I would love to see how much less CPU it uses then my current scheme :D 
https://www.youtube.com/watch?v=hOsDv3P1BA4

**Comment by panzergame** [comment]
@youle31 : I would like that you test to modify physics attributes when the object is disabled.
Like reinstancing a physics shape, change the mass, apply a velocity ect...

**Comment by youle31** [comment]
@BluePrintRandom : could you test this build: https://drive.google.com/file/d/0B3GouQIyoCmrZXBfTU01djNxM3M/view?usp=sharing

to see if this works for you and report us if you have bugs, performances gain etc...? Thanks!

**Comment by BluePrintRandom** [comment]
sure!

**Comment by BluePrintRandom** [comment]
object.suspendPhysics() and object.restorePhysics() right?

what is the api?

**Comment by BluePrintRandom** [comment]
added.suspendPhysics = True

got it I think

**Comment by BluePrintRandom** [comment]
Ok! tested - it's a little faster then adding/removing objects each frame,
and also it removes the complexity of needing a physics mesh and a seperate
graphics mesh for my kdtree example and also decreases the bloat to
scene.objects

so
benefits -
*less objects* = faster item lookup
*suspend is faster than *- add/remove
*reduced complexity - *Now it's easier to 'ghost' things as well


*http://pasteall.org/blend/index.php?id=43423
<http://pasteall.org/blend/index.php?id=43423>*

File: source/gameengine/Ketsji/KX_GameObject.cpp [hunk 1]
```diff
@@ -236,7 +240,10 @@ void KX_GameObject::SetName(const char *name)
 
 PHY_IPhysicsController* KX_GameObject::GetPhysicsController()
 {
-	return m_pPhysicsController;
+	if (m_pPhysicsController && !m_suspendedPhysics) {
```
Comment (panzergame) [inline source/gameengine/Ketsji/KX_GameObject.cpp]: These modifications are needed ?

File: source/gameengine/Ketsji/KX_GameObject.cpp [hunk 2]
```diff
@@ -120,7 +123,8 @@ KX_GameObject::KX_GameObject(
       m_components(NULL),
       m_pInstanceObjects(NULL),
       m_pDupliGroupObject(NULL),
-      m_actionManager(NULL)
+      m_actionManager(NULL),
+	  m_pSavedPhysicsController(NULL)
```
Comment (panzergame) [inline source/gameengine/Ketsji/KX_GameObject.cpp]: What's the goal of `m_pSavedPhysicsController` ? Please avoid the type in name convention : replace `m_p` by `m_`.

**Comment by panzergame** [comment]
Using the physics controller while it is suspended should not be a problem. The physics controller is still valid (it's rigid/soft body too) but just not used in the physics simulation world.

You just have to check that all the cases calling CcdPhysicsEnvironment::AddCcdPhysicsController are impossible when the physics controller is considerated suspended. I'm thinking about calling from python restoreDynamics.

File: source/gameengine/Ketsji/KX_GameObject.cpp (hunk 1 above)
Comment (youle31) [inline source/gameengine/Ketsji/KX_GameObject.cpp]: When we try to set an attribute if physics controller has been removed with : 
CcdPhysicsEnvironment *env = (CcdPhysicsEnvironment *)GetScene()->GetPhysicsEnvironment();
env->RemoveCcdPhysicsController((CcdPhysicsController *)GetPhysicsController());

it causes a crash.

File: source/gameengine/Ketsji/KX_GameObject.cpp (hunk 2 above)
Comment (youle31) [inline source/gameengine/Ketsji/KX_GameObject.cpp]: m_pSavedPhysicsController is not needed indeed

File: source/gameengine/Ketsji/KX_GameObject.cpp (hunk 1 above)
Comment (panzergame) [inline source/gameengine/Ketsji/KX_GameObject.cpp]: It happens for every physics attributes ?

**Comment by youle31** [comment]
When you call RemoveCcdPhysicsController you can't use anymore functions to set attributes like mass...

e.g: http://pasteall.org/blend/index.php?id=43454

File: source/gameengine/Ketsji/KX_GameObject.cpp (hunk 1 above)
Comment (youle31) [inline source/gameengine/Ketsji/KX_GameObject.cpp]: It happens at least for mass and as almost all attributes are set like mass, I guess this is the same

**Comment by BluePrintRandom** [comment]
Will this branch be merged soon?

(I like it :D)

**Comment by panzergame** [comment]
@youle31 : I propose you the patch : http://pasteall.org/82228/diff

I use `restorePhysics`and `disablePhysics` to be consistent with the functions `restoreDynamics` and `disableDynamics`.
The issue of changing mass with a suspended physics controller is resolved by checking if the physics controller is contained in the physics environment.
The same fix is applied for any functions using `UpdateCcdPhysicsController`in`CcdPhysicsController`.

**Comment by youle31** [comment]
ok cool :)

**Comment by panzergame** [comment]
@youle31 : Do you agree to apply my patch in the branch ?

@DCubix, @lordloki : Are you agree with the current works ? Can you test with the example and export it to bigger files for tests ?

**Comment by youle31** [comment]
Yes i agree. Or You Can do ot if You want

**Comment by youle31** [comment]
@panzergame : I can't apply your patch in ge_suspendphysics neither in master...

**Comment by panzergame** [comment]
Merged in master.

**Pull request closed without merging**